Upgrade CLI to use Click over Argparse

voetberg opened the pull request:
Resolves #7253 

Changes the argument parsing of CLI commands to use [click](https://click.palletsprojects.com/en/stable/) over argparse. Makes changes to accommodate this update, including adding `--no-x` options for boolean flags, removing documentation from positional args, implementing lazy loading for `rucio` commands, creating an `Arguments` type to facilitate translation between the original CLI using argparse and the new click CLI. 

Related changes: 
* Adds the suffix "commands" to all cli command files (avoid confusion with core files) 
* Updates the "command map" for directing users to the updated command 
* Replaces many positional args that were removed in the initial rewrite (#7250)
* Adds many `TODO`s where more work should be done

Additional issues potentially addressed: 
#7235 
#7309 
#7265 
#7301 
#7302 

Testing error is due to PyLint being upside the auth arguments are not being supplied to `commands.main()` when called in `bin/rucio`, `bin/rucio-admin`. An ignore statement is included there but it looks like it isn't considered. Additional guidance there would be appreciated.

voetberg committed:
Client: Replace argparse with click; Fix #7253

Changes:
	- Add the suffix "_command" to all cli files
	- Changed main CLI Parsing engine to click
	- Restored positional args from original cli (pre 36)
	- Replaced attribute changing args with --arg attribute value
	  formated arguments
	- Added '-' in args that should have had spaces
	- Reverted removing certian kwargs, opt to hide instead
	- Added TODO statements for future improvement possiblities

voetberg committed:
Client: Update commands in CLI Structure tests #7253

voetberg committed:
Clients: Update docstrings on commands #7253

voetberg committed:
Clients: Update mapping to new command for user hints #7253

voetberg committed:
Correct test failures

voetberg committed:
Clients: Mark NOQA on call to main click

voetberg committed:
Client: replace commands in test_rucio_server

voetberg committed:
Clients: Add Click as a client depedency #7253

voetberg committed:
Client: Update cli commands in tests

voetberg committed:
Delete lib/rucio/client/commands/rse.py in rebase

voetberg committed:
Delete lib/rucio/client/commands/rule.py in rebase

dchristidis commented:
> - Add the suffix "_command" to all cli files

Is this beneficial? On the server side, I always found it helpful that you expect to see the same filename in the REST API, gateway, and core layers.

But can we take a moment to plan the long-term structure? It appears that right now we have:

```
bin/rucio{,-admin}
lib/rucio/client/…
lib/rucio/client/commands/…
lib/rucio/client/commands/bin_legacy/…
```

`lib/rucio/client/…` likely has to stay as is. I applaud your choice to reduce `bin/rucio` to its simplest form, and have a new ‘commands’ pseudo-layer. I assume that the content of `lib/rucio/client/commands/bin_legacy/…` will eventually be moved to the ‘commands’? Meaning that both the command structure and their implementation will exist in the same functions?

But then the location of the `lib/rucio/client/commands/` directory looks somewhat out of place, and I don’t particularly like seeing, for example, `account_commands.py` and `accountclient.py`. Beyond the naming inconsistency, I think it creates a bit of confusion as to what is the CLI and what is the Client API layer. I wonder if `lib/rucio/cli/` might be preferable?

voetberg commented:
>> Add the suffix "_command" to all cli files

> Is this beneficial? On the server side, I always found it helpful that you expect to see the same filename in the REST API, gateway, and core layers.

I remember there being a conversation that everything being named the same was causing confusion, but I can no longer find written record of it, so I am just fine keeping them named without the `__commands` suffix if there are so objections. 

> I assume that the content of `lib/rucio/client/commands/bin_legacy/…` will eventually be moved to the ‘commands’? Meaning that both the command structure and their implementation will exist in the same functions?

Yes, that is the goal. I hope `bin_legacy` will cease to exist around 40.0.0. I have been having questions about this so I will make a small migration/upgrade plan on my fork for reference that we can iterate on. 

> I wonder if `lib/rucio/cli/` might be preferable?

I wouldn't have objections to moving the commands, but I don't have a strong reason to either. The changes were kept the `/client/` dir because they are all applications of client functionality and basically just shortcuts to combinations of client calls, so grouping them makes sense. I have absolutely 0 strong feelings on this though.

bari12 commented:
I had to reflect a bit about this, it's not a clear cut, but to me we did not do ourselves a favour with calling what is effectively the API, `clients`. This creates a cascading mess now, but within this mess, I don't think we should use the directory to also store the `commands`. I get the motivation, that everything which somebody might need in a client environment (API, CLI) is within the `lib/rucio/client` path. But what will happen, I think, is that people get confused and will use these commands as APIs and mixing it up with the "normal" clients (the actual API).

In the long run, I think we should separate this by moving `clients` to `api` and move this commands layer somewhere else. Could be `lib/rucio/cli`

voetberg commented:
I am not opposed to this at all, I can change this PR around to reflect that.

dchristidis commented on lib/rucio/client/commands/utils.py:
```diff
@@ -223,3 +214,21 @@ def setup_gfal2_logger():
+    @exception_handler
+    def func....
+    """
+    return exception_handler(click.pass_context(click.help_option("-h", "--help")(func)))
```
Help me out here, because I fail to see the need for this.

* `@click.help_option("-h", "--help")` can be replaced with `@click.group(context_settings={"help_option_names": ["-h", "--help"]})` above `main()`.
* `@click.pass_context` can be used where appropriate.
* `@exception_handler` can also be put above `main()` and it will apply to all subcommands.

The only thing that I didn’t try in my own experiments was lazy-loading.

voetberg replied:
Admittedly I didn't know about the `"help_option_names"` setting until after submitting this PR, and you're right about the other options as well

dchristidis commented on lib/rucio/client/commands/account_commands.py:
```diff
@@ -0,0 +1,219 @@
+@account.command("update")
+@click.argument("account-name")
+@click.option("--email", help="Email address associated with the account")
+@click.option("--ban/--unban", default=None, help="Temperately disable/enable an account")
```
Temperately?

voetberg replied:
"temporarily" 😶‍🌫️ Spell check does not have my best interests in mind it seems

dchristidis commented on lib/rucio/client/commands/account_commands.py:
```diff
@@ -0,0 +1,219 @@
+@click.argument("account-name")
+@click_decorator
+def attr_list(ctx, account_name):
+    "List the attributes for a given account. Note: This table is generally empty."
```
‘This table is generally empty’ is community-specific.

voetberg replied:
This was grabbed from the original list-attr in rucio-admin, but I will remove it.

dchristidis replied:
Very well, thank you for the explanation. Then I’ll try to ignore the actual help content as much as possible and focus solely on the structure and implementation.

dchristidis commented on lib/rucio/client/commands/account_commands.py:
```diff
@@ -0,0 +1,219 @@
+
+@attribute.command("add")
+@click.argument("account-name")
+@click.option("-o", "--option", type=(str, str), help="Key and value of the attribute", required=True)
```
This hinders the review process. You’re making a notable change to the command-line options that is hidden in a diff between a file that is entirely removed and a new file. I must ask you to make a commit that solely makes the use of Click without any other changes, unless absolutely mandatory (and those should be well-advertised in the commit message).

In this case, I find the change debatable. The user will need to remember that it’s `--option` to add but `--key` to remove. Personally, I don’t think it beneficial.

voetberg replied:
I will revert these and add them as a discussion topic.

voetberg replied:
View/add comments to https://github.com/rucio/rucio/issues/7523

dchristidis commented on lib/rucio/client/commands/account_commands.py:
```diff
@@ -0,0 +1,219 @@
+@click.argument(
+    "account-name",
+)
+@click.option("--rse", "--rse-name", help="Full RSE name", required=True)
```
This is a weird case. With `local` scope it’s and RSE name, but with `global` scope it’s an RSE expression.

voetberg replied:
Two possible solutions are presented in https://github.com/rucio/rucio/issues/7505 - I will add a todo linking to that issue

dchristidis commented on lib/rucio/client/commands/account_commands.py:
```diff
@@ -0,0 +1,219 @@
+@identity.command("list")
+@click.argument("account-name", required=True)
+@click_decorator
+def id_list(ctx, account_name):
```
Isn’t this naming scheme difficult to follow? Why not name it `account_identity_list()` so that it matches the actual command?

voetberg replied:
The `identity` commands were named with `id` so `id_add` would not conflict with the existing `identity_add` (from rucio-admin) - this should be solved with a `import as`. 

The general naming scheme is: 
* base group commands are named for their operation (`[group]_command.py::[operation]`)
* Subgroups are named for their subgroup and operation (`[group]_command.py::[subgroup]_[operation]`)

dchristidis replied:
I have my own thoughts, but this has no effect on the end user so rightfully the decision rests with you.

dchristidis commented on lib/rucio/client/commands/account_commands.py:
```diff
@@ -0,0 +1,219 @@
+@click.option("--type", "type_", type=click.Choice(["X509", "GSS", "USERPASS", "SSH", "SAML", "OIDC"]), help="Authentication type", required=True)
+@click.option("--id", help="Identity", required=True)
+@click.option("--email", help="Email address associated with the identity", required=True)
+@click.option("-pwd", "--password", help="Password if authtype is USERPASS")
```
Big no-no! There are three main conventions for command-line options: one-dash-one-letter (UNIX), one-dash-multiple-letters (X11), two-dashes-multiple-letters (GNU). The UNIX convention hash the added trick that options can be combined. Implementing all three in a utility sounds like a recipe for disaster (think about [tar](https://xkcd.com/1168/) and how unintuitive it is due to trying to support multiple conventions).

voetberg replied:
This was pulled from the main oauth commands (to keep argument names consistent). Should it be changed there as well?

dchristidis replied:
Wow, I can’t believe this existed since 2012. Yes, this should be removed (even the idea that a password is passed as a command-line argument is ludicrous to me).

dchristidis commented on lib/rucio/client/commands/command.py:
```diff
@@ -11,253 +11,252 @@
-    def _operations(self):
-        return {}
+# Taken directly from https://click.palletsprojects.com/en/stable/complex/#defining-the-lazy-group
+class LazyGroup(click.Group):
```
How does lazy-loading benefit us?

voetberg replied:
To keep the `command.py` file from growing to unwieldy levels - it is largely an organizational feature in this case as the client themselves are light-weight to load in. I will argue for keeping lazy-loading though, for developer quality of life (I personally do not want to work with a file that could end up being over 6000 lines long when bin_legacy is removed.)

dchristidis replied:
Thank you, no argument against it.

dchristidis commented on lib/rucio/client/commands/command.py:
```diff
@@ -11,253 +11,252 @@
+)  # TODO: Implement https://click.palletsprojects.com/en/stable/options/#dynamic-defaults-for-prompts for args from config or os
+@click.version_option(version.version_string(), message="%(prog)s %(version)s")
+@click.option("--config", help="The Rucio configuration file to use")
+@click.option("-v", "--verbose/--no-verbose", default=False, help="Print more verbose output")
```
I don’t see the need for having two options for a boolean when one of the is the default.

voetberg replied:
I think I was still figuring out how click handles bools here - I will replace it with a `is_flag` option.

dchristidis replied:
This applies to other options as well.

voetberg replied:
I will argue for using this in some cases (Having explicit true/false flags is useful when you know the option exists but can't remember the default - a not unheard-of situation), but in the spirit of not making changes explicitly outside the migration I will put it up as a discussion topic.

dchristidis replied:
But looking at how you’re using this feature, you end up in the opposite situation. You say ‘can’t remember the default’, instead I see ‘cant’ remember the prefix’.

```
@click.option("--locked/--unlocked", default=False, help="Rule locking")  # Kept so rules can be unlocked as well as locked
```

Before I would have had to remember that the option is called `--locked` and I can pass it either `True` or `False`. If I don’t remember what is the default, I can set to what I want anyway. With you suggestion, I have to remember than the opposite of `--locked` is `--unlocked`. And in the very last commit, some options were prefixed with `--no-` and some with `--not-`. Why do you think it more likely to not remember the default than the opposite?

Please revert that one as well and open an issue if you wish to pursue this further. I am but one voice and I should not be making such decisions. But that means that they should not be bundled with the argparse-to-click migration.

dchristidis commented on lib/rucio/client/commands/config_commands.py:
```diff
@@ -0,0 +1,74 @@
+@click.group()
+@click.help_option("-h", "--help")
+def config():
+    "Modify the rucio.cfg"
```
Not really. The modification applies to the config table in the database.

dchristidis commented on lib/rucio/client/commands/config_commands.py:
```diff
@@ -0,0 +1,74 @@
+@config.command("remove")
+@click.option("-s", "--section", help="Section", required=True)
+@click.option("-k", "--key", help="Key in section", required=True)
+@click.confirmation_option(prompt="Are you sure you want to delete this value?")
```
Again, this is a new feature. Please don’t hide it under the argparse-to-Click migration.

dchristidis commented on lib/rucio/client/commands/did_commands.py:
```diff
@@ -0,0 +1,192 @@
+def remove(ctx, dids, undo):
+    """
+    This command sets the lifetime of the DID in order to expire in the next 24 hours.
+    After this time, the dataset is eligible for deletion.
```
‘Eligible for deletion’ suggests that it is at someone’s discretion. Expired DIDs are force-deleted (and their replicas purged).

voetberg replied:
This was pulled from the original help menu, I will replace it with your explanation

dchristidis commented on lib/rucio/client/commands/lifetime_exception_commands.py:
```diff
@@ -0,0 +1,34 @@
+@click.option("-x", "--expiration", help="The expiration date format YYYY-MM-DD")
+@click_decorator
+def add_(ctx, input_file, reason, expiration):
+    """Add an exception to the lifetime model (to have protections against deletion from reaper daemons)"""
```
This description is very wrong. Where did it come from?

voetberg replied:
Honestly not sure. Removing it now.

dchristidis commented on lib/rucio/client/commands/rse_commands.py:
```diff
@@ -0,0 +1,305 @@
+@click.confirmation_option(prompt="Are you sure you want to delete this RSE?")
+@click_decorator
+def remove(ctx, rse_name):
+    """Permanently disable an RSE. WARNING: You may not be able to recover this RSE!"""
```
Again, what does ‘not be able to recover’ mean?

voetberg replied:
The original from rucio-admin states "CAUTION: all information about the RSE might be lost!" which I think got mutated into the current version. I'll replace the pr comment with that

dchristidis commented on lib/rucio/client/commands/rse_commands.py:
```diff
@@ -0,0 +1,305 @@
+def distance_update(ctx, source_rse, destination_rse, distance, ranking):
+    """Update the existing distance or ranking from SOURCE-RSE to DESTINATION-RSE"""
+    if (distance is None) and (ranking is None):
+        raise ValueError("Either [distance] or [ranking] must be specified")
```
Their default value is `1`. Under what circumstances will either be `None`?

Also why is ranking still referenced? It has been removed from the database.

voetberg replied:
It was not removed from the cli so it was replicated here (just a confusing string change warning is there). I will make that change

dchristidis replied:
See #5985 for reference. I’m not sure whether the client was ignored accidentally or on purpose, but it’s been long enough that we can remove it.

dchristidis commented on lib/rucio/client/commands/rule_commands.py:
```diff
@@ -0,0 +1,159 @@
+@click.option("--source-rses", help="Update how replicas are sourced for the rule")
+@click_decorator
+def move(ctx, rule_id, rses, activity, source_rses):
+    """Duplicate a rule onto another RSE"""  # TODO: verify - does it delete the old one?
```
No, it doesn’t duplicate. The two rules exist in parallel with a parent-child relationship. When the child rule reaches `OK` state, the parent is deleted.

voetberg committed:
Clients: Apply review comments #7253

Changes:
	* Remove extra 'help' menu signatures
	* Correct misleading comments and function discriptions
	* Remove the change of --key key --value value arguments from -o arguments
	* Make naming schema completely uniform

voetberg committed:
Clients: Rename client/commands to cli #7253

voetberg committed:
Clients: Disable Pylint error #7253

voetberg committed:
Clients: Add test verifying -h and --help help menus for commands and subcommands

voetberg committed:
Client: correct config storing, conflicting commands

voetberg committed:
Client: Correct errors with syntax

voetberg commented:
Remarks re: Failing tests 
* Type checks fail because of 2 things - incorrectly picking up `bin_legacy/rucio[-admin]` and not regarding the typing on arguments set by `click`. The first should be ignored, the second is open for discussion on if these arguments should have typing applied (double work, as `click` already sets the default type of `Union[str, None]` unless otherwise specified)
* The  Type Check PR also fails because of bin_legacy and should be ignored.

dchristidis commented on lib/rucio/cli/command.py:
```diff
@@ -0,0 +1,267 @@
+@click.option("-u", "--user", help="Username for userpass")
+@click.option("-pwd", "--password", help="Password for userpass")
+@click.option("--oidc-user", help="OIDC username")
+@click.option("--oidc-password", help="OIDC password")
```
Both argparse and Click preserve the option order. There’s no logic to their ordering, at least none that I can tell. Should we consider ordering them alphabetically (base on the long option)?

voetberg replied:
I think grouping user/pass options together is a good idea, but otherwise I don't see a problem with that

voetberg committed:
Client: sort main CLI options #7253

voetberg committed:
Client: Remove non-used ranking option from RSE CLI #7253

voetberg committed:
Client: Remove boolean -x/-no-x options #7253

dchristidis reviewed (commented):
There’s much in this PR that merit a closer look. A lot of it is cruft accumulated over multiple years of Rucio development (and not always with a clear long-term vision). Some is not, but I’m now asking myself whether this PR fixes what it set out to do. The response is a resounding **YES**.

Please address the only remaining open comment and rebase on top of the latest master. I am prepared to give my approval (and I am very eager to test the release candidate in action).

Well done.

dchristidis commented on lib/rucio/cli/upload.py:
```diff
@@ -0,0 +1,60 @@
+@click.option("--scope", help="Scope name.")
+@click.option("--impl", type=click.Choice([]), help="Transfer protocol implementation to use (e.g: xrootd, gfal.NoRename, webdav, ssh.Rsync, rclone)")
+# The --no-register option is hidden. This is pilot ONLY. Users should not use this. Will lead to unregistered data on storage!
+@click.option("--no-register", is_flag=True, default=False, hidden=True)
```
~~I think you need to add an additional `"register"` argument here for the tests to pass.~~

Rather rename the function parameter below to `no_register`.

voetberg replied:
🤦 Totally missed that one. Corrected in latest force push
